target/linux/ipq40xx/patches-5.4/200-arm-compressed-add-appended-DTB-section.patch

```diff
@@ -0,0 +1,48 @@
+From 0843a61d6913bdac8889eb048ed89f7903059787 Mon Sep 17 00:00:00 2001
+From: Robert Marko <robimarko@gmail.com>
+Date: Fri, 30 Oct 2020 13:36:31 +0100
+Subject: [PATCH] arm: compressed: add appended DTB section
+
+This adds a appended_dtb section to the ARM decompressor
+linker script.
+
+This allows using the existing ARM zImage appended DTB support for
+appending a DTB to the raw ELF kernel.
+
+Its size is set to 1MB max to match the zImage appended DTB size limit.
+
+To use it to pass the DTB to the kernel, objcopy is used:
+
+objcopy --set-section-flags=.appended_dtb=alloc,contents \
+	--update-section=.appended_dtb=<target>.dtb vmlinux
+
+This is based off the following patch:
+https://github.com/openwrt/openwrt/commit/c063e27e02a9dcac0e7f5877fb154e58fa3e1a69
+
+Signed-off-by: Robert Marko <robimarko@gmail.com>
+---
+ arch/arm/boot/compressed/vmlinux.lds.S | 9 ++++++++-
+ 1 file changed, 8 insertions(+), 1 deletion(-)
+
+--- a/arch/arm/boot/compressed/vmlinux.lds.S
++++ b/arch/arm/boot/compressed/vmlinux.lds.S
+@@ -93,6 +93,13 @@ SECTIONS
+ 
+   _edata = .;
+ 
++  .appended_dtb : {
++    /* leave space for appended DTB */
++    . += 0x100000;
++  }
```

Comment:
This wants a wrapping in a Kconfig pre-processor conditional that is only set for our subtarget? Or testing on zImage ipq40xx devices, maybe Device/mobipromo_cm520-79f, or a linksys ip40xx device, though these only use zImage indirectly?
Also, I think 301- would be a better suit for the patch number.

Reply:
This should work fine for zImage devices as well as it at worst simply leave empty space.
But, I can add a Kconfig symbol and enable it for MikroTik only.

Yeah, will change the numbering.

Reply:
Will not need Kconfig precursor for this patch:
I tried this patch with Openwrt's armvirt 32 zImage, and it did not cause issue:
- armvirt 32 (QEMU's DTB) okay:
 `qemu-system-arm -nographic -M virt -m 64 -kernel openwrt-armvirt-32-zImage-initramfs`
- armvirt 32 without DTB: no boot, as expected:
  create empty dtb for qemu:
  ```
  cat <<-EOF |   dtc -I dts -O dtb -o none.dtb
  /dts-v1/;

  / {
    #address-cells = <1>;
    #size-cells = <1>;
  };
  EOF

  qemu-system-arm -nographic -M virt -m 64 -dtb none.dtb -kernel openwrt-armvirt-32-zImage-initramfs
  ```
- armvirt 32 with `CONFIG_ARM_APPENDED_DTB=y`: Boots (with QEMU DTB, or zImage appended DTB) either (with or without this patch):
  ```
  # extract qemu dtb
  qemu-system-arm -nographic -M virt -m 64 -machine dumpdtb=qemu_arm_virt.dtb
  # append it to the zImage
  cat openwrt-armvirt-32-zImage-initramfs qemu_arm_virt.dtb > openwrt-armvirt-32-zImage-initramfs-appended-dtb-wdtb
  # boot with QEMU
  qemu-system-arm -nographic -M virt -m 64 -dtb none.dtb -kernel openwrt-armvirt-32-zImage-initramfs-appended-dtb-wdtb
  ```
  DEBUG_LL settings for QEMU armvirt:
  ```
  CONFIG_DEBUG_LL=y
  CONFIG_DEBUG_LL_INCLUDE="debug/pl01x.S"
  CONFIG_DEBUG_LL_UART_PL01X=y
  CONFIG_DEBUG_UART_PHYS=9000000
  CONFIG_DEBUG_UART_VIRT=c9000000
  CONFIG_DEBUG_UNCOMPRESS=y
  CONFIG_EARLY_PRINTK=y
  ```
  Needs earlyprintk linux command line:
  `qemu-system-arm -nographic -M virt -m 64 -dtb none.dtb -append "earlyprintk=/dev/ttyAMA0" -kernel openwrt-armvirt-32-zImage-initramfs-appended-dtb-wdtb`

Could not get a linux kernel ELF to boot for me in arm qemu, but happy enough to say it seems we are not butchering zImage with these changes.

Cheers

Reply:
Great, I have finally got around to rebasing this today.